### ISSUE TITLE
User reports: UnicodeDecodeError: 'ascii' codec can't decode byte

### ISSUE THREAD

**Issue by bwbeach** [Backblaze/B2_Command_Line_Tool#190]
```
Traceback (most recent call last): 
File "/usr/local/bin/b2", line 9, in <module> 
load_entry_point('b2==0.5.7', 'console_scripts', 'b2')() 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/console_tool.py", line 870, in main 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/console_tool.py", line 797, in run_command 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/console_tool.py", line 621, in run 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 912, in sync_folders 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 812, in make_folder_sync_actions 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 653, in zip_folders 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 622, in next_or_none 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 639, in <genexpr> 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 498, in all_files 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 546, in _walk_relative_paths 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 546, in _walk_relative_paths 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 546, in _walk_relative_paths 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 535, in _walk_relative_paths 
File "/usr/local/lib/python2.7/posixpath.py", line 73, in join 
path += '/' + b 
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 51: ordinal not in range(128)
```

**Comment by bwbeach** [comment]
The stack trace shows the latest dev version number, and python 2.7.

After experimenting with things, the only way I could get this error from concatenating strings is by having one str and one unicode:

```
>>> '\xff' + u'foo'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)
```

**Comment by bwbeach** [comment]
`os.listdir` should be returning unicode, not str.  I don't see how this could happen.

**Comment by bwbeach** [comment]
The master branch, and the released version 0.6.0 both have an assertion to help diagnose this problem.  If you're seeing this, could you try the latest and see what happens.

**Comment by nilayp** [comment]
A customer has seen this assertion in the field. Here is their stack trace:

`Traceback (most recent call last): 
File "/usr/local/bin/b2", line 9, in <module> 
load_entry_point('b2==0.6.3', 'console_scripts', 'b2')() 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/console_tool.py", line 879, in main 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/console_tool.py", line 806, in run_command 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/console_tool.py", line 623, in run 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 951, in sync_folders 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 848, in make_folder_sync_actions 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 696, in zip_folders 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 656, in next_or_none 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 674, in <genexpr> 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 509, in all_files 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 580, in _walk_relative_paths 
File "build/bdist.freebsd-10.3-STABLE-amd64/egg/b2/sync.py", line 551, in _walk_relative_paths 
AssertionError: non-ascii character in str returned by os.listdir: 239`

**Comment by clintinthecode** [comment]
Hey guys, I am the one who submitted the bug via email support. If there is any further info I can provide, or any testing we can do to help resolve the issue ASAP please let me know. Happy to help.

**Comment by ppolewicz** [comment]
@clintinthecode, please run this in the same environment in which you have encountered the problem and send us back the resulting output:
```python
from __future__ import print_function
import sys
import platform
print(sys.version)
print(platform.platform())
print(platform.machine())
print(platform.processor())
```
It could also help if you'd tell us what is the exact strange path you are trying to sync and how can we make such path in our environment (unless it is your secret).

**Comment by clintinthecode** [comment]
Hi @ppolewicz, The output is below:

2.7.11 (default, Mar 19 2016, 01:13:51) 
[GCC 4.2.1 Compatible FreeBSD Clang 3.4.1 (tags/RELEASE_34/dot1-final 208032)]
FreeBSD-10.3-STABLE-amd64-64bit-ELF
amd64
amd64

The machine is an intel server in our Sydney office, supermicro motherboard, intel Xeon processor, purchased all the hardware a month ago, the OS is FreeNAS, running the latest stable version, I have a jail running which is where the B2 Command Line Tool and the code above was executed from.

I did give a list of filenames to the Backblaze support team but I guess it didn't get passed on.

directory “Mp3 ” the last character there on my mac is a space, via terminal its a square with a question mark
directory “Intro/outro” on my mac its a / but on terminal its a square with a question mark
file “How to ‘Get Off The Tools’ by Developing Business Systems” the single quotes in terminal don’t work but on my mac they are a quote at a bit of an angle different to the normal ‘ symbol
file “Öberg” The Ö doesn’t display in terminal but does in mac.

.... don't blame me for the / in the folder name Intro/Outro lol... Thats the creative people in the office responsible for that.. Its used frequently in filenames.

I spent half a day searching through the 14 Terrabytes of files and folders renaming everything to get rid of the non ASCII characters, so the backup is currently running, but I know there will be non ASCII characters used again in the future.

**Comment by ppolewicz** [comment]
@clintinthecode thank you for the detailed explanation.
If you take a file with a weird name like that and try to sync it from your Mac or a Linux computer, does the same problem occur?

**Comment by bwbeach** [comment]
The error report is coming from code that calls os.listdir().  It expects that when passing it a unicode string, it will get unicode strings in the response.

This is the code in posixmodule.c from Python 2.7 that decides how to return the results of os.listdir():

    #ifdef Py_USING_UNICODE
        if (arg_is_unicode) {
            PyObject *w;

            w = PyUnicode_FromEncodedObject(v,
                                            Py_FileSystemDefaultEncoding,
                                            "strict");
            if (w != NULL) {
                Py_DECREF(v);
                v = w;
            }
            else {
                /* fall back to the original byte string, as
                   discussed in patch #683592 */
                PyErr_Clear();
            }
        }
    #endif

It looks like the a str will be returned if the file name can't be decoded using the default file system encoding.  I wonder if the problem is showing up when the default file system encoding is set wrong.

**Comment by bwbeach** [comment]
@clintinthecode: Thanks for all your help on this.  I don't know how much time you have, but I would appreciate a couple more bits of info.

Could make a file with a unicode name:

    mkdir tmp
    cd tmp
    echo hello > 自由.txt

and then try this?

    import os
    import sys
    print sys.getfilesystemencoding()
    print os.listdir('.')
    print os.listdir(u'.')

This is what it does on my Mac:

```
Python 2.7.11 (default, Mar  7 2016, 13:29:38) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> print sys.getfilesystemencoding()
utf-8
>>> import os
>>> print os.listdir('.')
['\xe8\x87\xaa\xe7\x94\xb1.txt']
>>> print os.listdir(u'.')
[u'\u81ea\u7531.txt']
```

**Comment by bwbeach** [comment]
And one more thing would be useful: calling `os.listdir()` in a directory that has a file that fails to upload:

```
import os
print os.listdir('.')
print os.listdir(u'.')
```

**Comment by clintinthecode** [comment]
I wasn't able to directly create a file with that name, i couldn't copy and paste the 自由 characters into terminal on my mac (SSH'd into the BSD Server). However I do have a mapped network drive on my mac to a directory on the server, so i used my mac to create the text file with that name, and  then copy th file into the tmp dir.

Interestingly when using tab complete it showed the filename with the characters as expected, however when doing ls on that dir it instead says ??????.txt

[root@Backup_Service_Server ~/tmp]# cp /mnt/Videos/clint\ temp/
.DS_Store   自由.txt  
[root@Backup_Service_Server ~/tmp]# cp /mnt/Videos/clint\ temp/* .
[root@Backup_Service_Server ~/tmp]# ls
.txt		??????.txt
[root@Backup_Service_Server ~/tmp]# ls /mnt/Videos/clint\ temp/
.DS_Store	??????.txt


As requested I executed the following code, and the result is below. Im creating a new file test.py, putting the code in that file and running python test.py to get the results below.

**test.py**
import os
import sys
print sys.getfilesystemencoding()
print os.listdir('.')
print os.listdir(u'.')

**Result of execution**
US-ASCII
['\xe8\x87\xaa\xe7\x94\xb1.txt', 'test.py']
['\xe8\x87\xaa\xe7\x94\xb1.txt', u'test.py']

**Directory listing using ls**
[root@Backup_Service_Server ~/tmp]# ls
test.py		??????.txt



And for the second request..
import os
print os.listdir('.')
print os.listdir(u'.')

**Result of execution**
['test.py', 'test \xc3\x96berg.txt']
[u'test.py', 'test \xc3\x96berg.txt']

**Directory listing using ls**
[root@Backup_Service_Server ~/tmp]# ls
test ??berg.txt	test.py

**Dir listing during tab complete**
test Öberg.txt  test.py

**Comment by clintinthecode** [comment]
Interestingly I can paste 自由 into terminal on my mac at the prompt:
backupmacserver:~ clint$ 自由 

however pasting that into the terminal on my mac SSH'd into the bsd server, doing the exact same thing, having an empty command prompt, then pressing command v to paste the two characters into terminal, i get the following as soon as i press command v. Im guessing it just cant handle those characters?


[root@Backup_Service_Server ~/Oberg]# \! . 2to3 2to3-2 2to3-2.7 \: CC IPXrouted Mail \[ \[\[ ]] a2p ac accton acpiconf acpidb acpidump addftinfo addr2line adduser adjkerntz afmtodit alias amd amq ancontrol aot-compile48 apm apply apr-1-config apropos apu-1-config ar arp as asa asf asn1_compile at ath3kfw atmconfig atq atrm audit auditd auditdistd auditreduce authpf authpf-noip automount automountd autounmountd awk b2 b64decode b64encode badsect banner basename bash bashbug batch bc bcd bcmfw bdes bg bhyve bhyvectl bhyveload biff bind binmiscctl boot0cfg bootparamd bootpef bootptest brandelf break bsdconfig bsdcpio bsdgrep bsdiff bsdinstall bsdlabel bsdtar bsnmpd bsnmpget bsnmpset bsnmpwalk bspatch bt3cfw bthidcontrol bthidd bthost btpand btsockstat btxld builtin bunzip2 byacc bzcat bzegrep bzfgrep bzgrep bzip2 bzip2recover bzless c++ c++48 c++filt c2ph c89 c99 caesar cal calendar callbootd caller camcontrol camdd cap_mkdb case cat catman cc ccdconfig cd cdcontrol chat checknr chflags chfn chgrp chio chkey chkgrp chkprintcap chmod chown chpass chroot chsh ci ckdist cksum clang clang++ clang-cpp clang-tblgen clear clear_locks clri cmp co col colcrt colldef colrm column comcontrol comm command compgen compile_et complete compopt compress config config_data conscontrol continue coproc corelist cp cpan cpasswd cpio cpp cpp48 cpucontrol cpuset crashinfo cron crontab crunchgen crunchide crypt csh csplit csup ctags ctfconvert ctfdump ctfmerge ctladm ctld ctlstat ctm ctm_dequeue ctm_rmail ctm_smail cu curl curl-config cut cvsps daemon date db_archive-5 db_archive-5.3 db_checkpoint-5 db_checkpoint-5.3 db_deadlock-5 db_deadlock-5.3 db_dump-5 db_dump-5.3 db_dump185-5 db_dump185-5.3 db_hotbackup-5 db_hotbackup-5.3 db_load-5 db_load-5.3 db_log_verify-5 db_log_verify-5.3 db_printlog-5 db_printlog-5.3 db_recover-5 db_recover-5.3 db_replicate-5 db_replicate-5.3 db_stat-5 db_stat-5.3 db_tuner-5 db_tuner-5.3 db_upgrade-5 db_upgrade-5.3 db_verify-5 db_verify-5.3 dc dconschat dd ddb declare devctl devd devfs devinfo df dhclient dhclient-script dialog diff diff3 digictl dirname dirs diskinfo disklabel disown dmesg do domainname done dpv drill dtc dtrace dtruss du dump dumpcis dumpfs dumpon dwp easy_install easy_install-2.7 echo ed edit editmap edquota ee egrep elf2aout elfdump elfedit elif else enable enc2xs enigma env envsubst eqn esac etcupdate etherswitchcfg eval evim ex exec exit expand export expr extattrctl f2py f2py-2.7 factor faithd false fastboot fasthalt fc fdcontrol fdformat fdisk fdread fdwrite fetch ffsinfo fg fgrep fi fifolog_create fifolog_reader fifolog_writer file file2c find find2perl finger fixmount flex flex++ flowctl fmt fmtree fold for fortune freebsd-update freebsd-version from fsck fsck_4.2bsd fsck_ffs fsck_msdosfs fsck_ufs fsdb fsfs-access-map fsinfo fsirand fstat fstyp fsync ftp ftp-proxy function fuser fwcontrol g++ g++48 gappletviewer48 gate-ftp gbde gc-analyze48 gcache gcc gcc-ar48 gcc-nm48 gcc-ranlib48 gcc48 gcj-dbtool48 gcj48 gcjh48 gconcat gcore gcov48 gdb gdbm_dump gdbm_load gdbmtool gdbserver gdbtui geli gencat gensnmptree geom getconf getent getextattr getfacl getfmac getopt getopts getpmac gettext gettext.sh gfortran gfortran48 ggatec ggated ggatel gij48 git git-cvsserver git-p4.py git-receive-pack git-shell git-upload-archive git-upload-pack gjar48 gjarsigner48 gjavah48 gjournal gkeytool48 glabel gmirror gmountver gmul(arg: 1) \! . 2to3 2to3-2 2to3-2.7 \: CC IPXrouted Mail \[ \[\[ ]] a2p ac accton acpiconf acpidb acpidump addftinfo addr2line adduser aojkerntz afmtodit alias amd amq ancontrol aot-compile48 apm apply apr-1-config apropos apu-1-config ar arp as asa asf asn1_compile at ath3kfw atmconfig atq atrm audit auditd auditdistd auditreduce authpf authpf-noip automount automountd autounmountd awk b2 b64decode b64encode badsect banner basename bash bashbug batch bc bcd bcmfw bdes bg bhyve bhyvectl bhyveload biff bind binmiscctl boot0cfg bootparamd bootpef bootptest brandelf break bsdconfig bsdcpio bsdgrep bsdiff bsdinstall bsdlabel bsdtar bsnmpd bsnmpget bsnmpset bsnmpwalk bspatch bt3cfw bthidcontrol bthidd bthost btpand btsockstat btxld builtin bunzip2 byacc bzcat bzegrep bzfgrep bzgrep bzip2 bzip2recover bzless c++ c++48 c++filt c2ph c89 c99 caesar cal calendar callbootd caller camcontrol camdd cap_mkdb case cat catman cc ccdconfig cd cdcontrol chxt checknr chflags chfn chgrp chio chkey chkgrp chkprintcap chmod chown chpass chroot chsh ci ckdist cksum clang clang++ clang-cpp clang-tblgen clear clear_locks clri cmp co col colcrt colldef colrm column comcontrol comm command compgen compile_et complete compopt compress config config_data conscontrol continue coproc corelist cp cpan cpasswd cpio cpp cpp48 cpucontrol cpuset crashinfo cron crontab crunchgen crunchide crypt csh csplit csup ctags ctfconvert ctfdump ctfmerge ctladm ctld ctlstat ctm ctm_dequeue ctm_rmail ctm_smail cu curl cmrl-config cut cvsps daemon date db_archive-5 db_archive-5.3 db_checkpoint-5 db_checkpoint-5.3 db_deadlock-5 db_deadlock-5.3 db_dump-5 db_dump-5.3 db_dump185-5 db_dump185-5.3 db_hotbackup-5 db_hotbackup-5.3 db_load-5 db_load-5.3 db_log_verify-5 db_log_verify-5.3 db_printlog-5 db_printlog-5.3 db_recover-5 db_recover-5.3 db_replicate-5 db_replicate-5.3 db_stat-5 db_stat-5.3 db_tuner-5 db_tuner-5.3 db_upgrade-5 db_upgrade-5.3 db_verify-5 db_verify-5.3 dc dconschat dd ddb declare devctl devd devfs devinfo df dhclient dhclient-script dialog diff diff3 digictl dirname dirs diskinfo disklabel disown dmesg do domainname done dpv drill dtc dtrace dtruss du dump dumpcis dumpfs dumpon dwp easy_install easy_install-2.7 echo ed edit editmap edquota ee egrep elf2aout elfdump elfedit elif else enable enc2xs enigma env envsubst eqn esac etcupdate etherswitchcfg eval evim ex exec exit expand export expr extattrctl f2py f2py-2.7 factor faithd false fastboot fasthalt fc fdcontrol fdformat fdisk fdread fdwrite fetch ffsinfo fg fgrep fi fifolog_create fifolog_reader fifolog_writer file file2c find find2perl finger fixmount flex flex++ flowctl fmt fmtree fold for fortune freebsd-update freebsd-version from fsck fsck_4.2bsd fsck_ffs fsck_msdosfs fsck_ufs fsdb fsfs-access-map fsinfo fsirand fstat fstyp fsync ftp ftp-proxy function fuser fwcontrol g++ g++48 gappletviewer48 gate-ftp gbde gc-analyze48 gcache gcc gcc-ar48 gcc-nm48 gcc-ranlib48 gcc48 gcj-dbtool48 gcj48 gcjh48 gconcat gcore gcov48 gdb gdbm_dump gdbm_load gdbmtool gdbserver gdbtui geli gencat gensnmptree geom getconf getent getextattr getfacl getfmac getopt getopts getpmac gettext gettext.sh gfortran gfortran48 ggatec ggated ggatel gij48 git git-cvsserver git-p4.py git-receive-pack git-shell git-uploadfarchive git-upload-pack gjar48 gjarsigner48 gjavah48 gjournal gkeytool48 glabel gmirror gmountver gmultipath gnative2ascii48 gnop gorbd48 gpart gperf gpioctl gprof graid graid3 grdc grep grmic48 grmid48 grmiregistry48 grn grodvi groff grog grolbp grolj4 grops grotty groups growfs gsched gserialver48 gshsec gssd gstat gstripe gtnameserv48 gunzip gvinum gvirstor gzcat gzexe gzip h2ph h2xs halt hash hastctl hastd hccontrol hcsecd hcseriald hd head help hexdump history hlfsd host hostapd hostapd_cli hostname hoststat hpftodit hv_kvp_daemon hxtool i2c iasl iconv id ident idle idle2 idle2.7 idprio if ifconfig ifmcstat in indent indexinfo indxbib inetd info infokey init install install-info instmodsh iostat ip6addrctl ipcrm ipcs ipf ipfs ipfstat ipftest ipfw ipfwpcap ipmon ipnat ippool ipresend iprop-log iscontrol iscsictl iscsid jail jcf-dump48 jexec jls jobs join jot json_pp jv-convert48 kadmin kbdcontrol kbdmap kcc kdestroy kdump kenv keylogin keylogout keyserv kf kgdb kgetcred kgmon kill killall kinit kldconfig kldload kldstat kldunload kldxref klist kpasswd krb5-config kstash ksu kswitch ktrace ktrdump ktutil l2control l2ping lam last lastcomm lastlogin ld ld.bfd ld.gold ldconfig ldd ldd32 leave less lessecho lesskey lesspipe.sh let lex lex++ libnetcfg limits link lint lkbib lmcconfig ln local local-unbound-setup locale locate lock lockf lockstat logger login logins logname logout look lookbib lorder lp lpc lpd lpq lpr lprm lptcontrol lptest ls lsextattr lsvfs lzcat lzdec lzegrep lzfgrep lzgrep lzless lzma lzmainfo m4 mail mailq mailstats mailwrapper mailx make make-roken makefs makeinfo makemap makewhatis man manctl mandoc manpath mapfile md5 mdconfig mdmfs memcontrol merge mergemaster mesg mfiutil minigzip ministat mixer mk-amd-map mkcsmapper mkdep mkdir mkesdb mkfifo mkimg mklocale mknod mksnap_ffs mkstr mktemp mkulzma mkuzip mld6query mlxcontrol mmroff more morse mount mount_cd9660 mount_fusefs mount_mfs mount_msdosfs mount_nfs mount_nullfs mount_oldnfs mount_smbfs mount_udf mount_unionfs mountd moused mptable mptutil msgs mt mtest mtree mv nano natd nawk nc ncal ndis_events ndiscvt ndisgen ndp neqn netstat newaliases newfs newfs_msdos newgrp newkey newsyslog nex nextboot nfscbd nfsd nfsdumpstate nfsiod nfsrevoke nfsstat nfsuserd ngctl ngettext nghook nice nl nm nmtree nohup nologin nos-tun nosetests nosetests-2.7 nroff nscd ntp-keygen ntpd ntpdate ntpdc ntpq ntptime number nvi nview nvmecontrol objcopy objdump od openssl opieinfo opiekey opiepasswd otp-md4 otp-md5 otp-sha1 pac pagesize passwd paste patch pathchk pawd pax pbi-crashhandler pbi_add pbi_addrepo pbi_autobuild pbi_browser pbi_create pbi_delete pbi_deleterepo pbi_icon pbi_indextool pbi_info pbi_listrepo pbi_makepatch pbi_makeport pbi_makerepo pbi_metatool pbi_patch pbi_update pbi_update_hashdir pbid pc-sysinstall pciconf periodic perl perl5 perl5.20.3 perlbug perldoc perlivp perlthanks perror pfbtops pfctl pflogd pftp pgrep pic piconv ping ping6 pkg pkg-static pkg2ng pkill pl2pm pmcannotate pmccontrol pmcstat pmcstudy pod2html pod2man pod2text pod2usage podchecker podselect pom popd portsnap post-grohtml powerd poweroff ppp pppctl ppt pr praliases praudit pre-grohtml primes printenv printf procctl procstat protect prove ps psed psroff pstat pstruct ptar ptardiff ptargrep purgestat pushd pw pwait pwd pwd_mkdb pydoc pydoc2 pydoc2.7 pyflakes pyflakes-2.7 python python-config python2 python2-config python2.7 python2.7-config quot quota quotacheck quotaoff quotaon random ranlib rarpd rbash rcorder rcp rcs rcsclean rcsdiff rcsfreeze rcsmerge rctl rdump read readarray readelf readlink readonly realpath reboot recoverdisk red ree refer renice repquota reset resolvconf restore return rev revoke rfcomm_pppd rfcomm_sppd rip6query rlog rlogin rm rmail rmd160 rmdir rmextattr rmt rmuser rnano rot13 route route6d routed rpc.lockd rpc.statd rpc.umntall rpc.yppasswdd rpc.ypupdated rpc.ypxfrd rpcbind rpcgen rpcinfo rrenumd rrestore rs rsh rtadvctl rtadvd rtprio rtquery rtsol rtsold rup ruptime rusers rview rvim rwall rwho rwhod s2p sa sade savecore scp script sdiff sdpcontrol sdpd sed select send-pr sendmail seq service services_mkdb sesutil set setextattr setfacl setfib setfmac setfsmac setkey setpmac sftp sh sha1 sha256 sha512 shar shasum shift shopt showmount shutdown sicontrol size slc sleep slogin smbmsg smbutil snapinfo sntp sockstat soelim sort source spkrtest splain split spppcontrol spray sqlite3 sscop ssh ssh-add ssh-agent ssh-copy-id ssh-keygen ssh-keyscan sshd stat stdbuf strfile string2key strings strip stty su sum suspend svn svn-populate-node-origins-index svnadmin svnauthz svnauthz-validate svnbench svndiff svndiff3 svndiff4 svndumpfilter svnfsfs svnlite svnliteadmin svnlitedumpfilter svnlitelook svnlitemucc svnliterdump svnliteserve svnlitesync svnliteversion svnlook svnmucc svnraisetreeconflict svnrdump svnserve svnsync svnversion swapctl swapinfo swapoff swapon sync sysctl syslogd sysrc systat tabs tail talk tar tbl tblgen tcopy tcpdchk tcpdmatch tcpdrop tcpdump tcsh tee telnet test texindex tfmtodit tftp then time timed timedc timeout times tip top touch tput tqdm tr traceroute traceroute6 trap troff trpt true truncate truss tset tsort tty tunefs type typeset tzsetup uathload uefisign ugidfw uhsoctl ul ulimit umask umount unalias uname unbound unbound-anchor unbound-checkconf unbound-control uncompress unexpand unifdef unifdefall uniq units unlink unlzma unset unstr until unvis unxz unzip uptime usbconfig usbdump usbhidaction usbhidctl users utx utxrm uudecode uuencode uuidgen vacation verify_krb5_conf vgrick zic zinject zipdetails zless zmore znew zpool zstreamdump ztest zzz \{ } testp zforce zfs zgrep zha

**Comment by bwbeach** [comment]
Thanks.  That info is very helpful.

It looks like the problem is that Python thinks the file system encoding is 'US-ASCII', when it fact it's UTF-8.

Digging in a bit, it looks like Mac and Windows have a standard encoding that Python knows about.  I test mostly on Mac, so it always works for me:

```
/* The default encoding used by the platform file system APIs
   Can remain NULL for all platforms that don't have such a concept
*/
#if defined(MS_WINDOWS) && defined(HAVE_USABLE_WCHAR_T)
const char *Py_FileSystemDefaultEncoding = "mbcs";
#elif defined(__APPLE__)
const char *Py_FileSystemDefaultEncoding = "utf-8";
#else
const char *Py_FileSystemDefaultEncoding = NULL; /* use default */
#endif
```

Testing on Linux, I find that setting the LANG environment variable will tell Python what encoding to use by default.  Try setting LANG to `en_US.UTF-8`, and see if that fixes the problem.  This is what happens on Debian:

```
$ unset LANG
$ python
Python 2.7.3 (default, Jun 21 2016, 18:38:19) 
[GCC 4.7.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; sys.getfilesystemencoding()
'ANSI_X3.4-1968'
>>> 
$ export LANG=en_US.UTF-8
$ python
Python 2.7.3 (default, Jun 21 2016, 18:38:19) 
[GCC 4.7.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys; sys.getfilesystemencoding()
'UTF-8'
>>> 
```

**Comment by ppolewicz** [comment]
I added the workaround to the error message, so that the user can spot it instead of an assertion.

@bwbeach: I think we should catch `UnicodeDecodeError` fired from `_walk_relative_paths` and then fire the new exception instead of manually walking through every single byte of every single filename on all platforms, just so that we can present a slightly nicer error message if the user has a mis-configured environment.

**Comment by clintinthecode** [comment]
That would be great if it could give a warning instead of crashing. I could then have our script spot the warning and email out mentioning the issue and mention the file in question.

Myself and one of my developers haven't been able to set the lang of the system successfully yet. We have followed a bunch of websites saying how to do it on Debian, but none of them have worked yet. After logging out and back in it always changes back to the wrong lang setting.

The workaround we did was to add a line to our script that sets the lang before calling sync. It works but isn't ideal.

Unless we have missed something, it doesn't always seem to be simple to set the lang, so others may encounter this issue.
